### PR TITLE
Fix rule set loading issue

### PR DESCRIPTION
**Pull request by derekgengenbacher-wf**
### Changes:
- Updated the runXule function in __init__.py to stop an error from occurring when an appropriate rule set does not exist in the rulesetMap. Prior to this change the determine_rule_set function in XuleUtilty.py could return None. Xule would then pass the None value to arelle for loading and arelle would error with an attribute error. The code now checks for the existence of a rule_set_location prior to passing it to arelle for loading. If no rule set exists then XULE short circuits and stops running.
- Cleaned up the determine_rule_set function in XuleUtility.py.
- Cleaned up the rulesetMap.json to no longer have references to rulesets that dont exist.

### Pull request process

- [ ] Full description of changes provided above.
- [ ] Reviewed by another person. (+1)
- [ ] If this is a code change, reviewed by a second person. (+1)
- [ ] QA'd to the specifications listed [here](https://github.com/DataQualityCommittee/dqc_us_rules#quality-assurance-qa-of-a-pull-request). Documentation only changes do not require the formal code quality assurance process. (+10)

##### Once checklist is complete, @campbellpryde @marcward @davidtauriello please review for merge.

### CODE REVIEW

**Comment by davidtauriello** [comment]
@derekgengenbacher-wf - thanks for submitting this pull; I'm moving it to v11 and hoping to get it QA'd for the next release

**Comment by derekgengenbacher-wf** [comment]
Great! Thank you David!!

**Comment by sagesmith-wf** [comment]
@Workiva/xt

File: plugin/xule/__init__.py [hunk 1]
```diff
@@ -775,52 +775,49 @@ def xuleCmdXbrlLoaded(cntlr, options, modelXbrl, *args, **kwargs):
...
+    if rule_set is None:
+        return
+
+    if getattr(options, "xule_multi", False):
+        xm.start_process(rule_set, modelXbrl, cntlr, options)
+    else:
+        if modelXbrl is None:
+            # check if there are any rules that need a model
+            for rule in rule_set.catalog['rules'].values():
+                if rule['dependencies']['instance'] == True and rule['dependencies']['rules-taxonomy'] != False:
```
Comment (sagesmith-wf) [inline plugin/xule/__init__.py]: Do these need to be `rule.get('dependencies', {}).get('instance', False)` and `rule.get('dependencies', {}).get('instance', None)`?

File: plugin/xule/__init__.py [hunk 2]
```diff
@@ -775,52 +775,49 @@ def xuleCmdXbrlLoaded(cntlr, options, modelXbrl, *args, **kwargs):
...
+                    raise xr.XuleRuleSetError('Need instance to process rules')
+
+        global _saved_taxonomies
+        used_taxonomies = process_xule(rule_set, modelXbrl, cntlr, options, _saved_taxonomies)
+        # Save one loaded taxonomy
+        new_taxonomy_keys = used_taxonomies.keys() - _saved_taxonomies.keys()
+        if len(new_taxonomy_keys) > 0:
```
Comment (sagesmith-wf) [inline plugin/xule/__init__.py]: Nit: `len(new_taxonomy_keys) >= 2`

File: plugin/xule/__init__.py [hunk 3]
```diff
@@ -775,52 +775,49 @@ def xuleCmdXbrlLoaded(cntlr, options, modelXbrl, *args, **kwargs):
...
-        except xr.XuleRuleSetError:
-            raise
-
-        if getattr(options, "xule_multi", False):
-            xm.start_process(rule_set,
-                         modelXbrl,
-                         cntlr,
-                         options
-                         )
+    if getattr(options, "xule_multi", True) and getattr(cntlr, "rule_set", None) is not None:
```
Comment (sagesmith-wf) [inline plugin/xule/__init__.py]: Should this default to `False` instead of `True`?

File: plugin/xule/__init__.py [hunk 4]
```diff
@@ -775,52 +775,49 @@ def xuleCmdXbrlLoaded(cntlr, options, modelXbrl, *args, **kwargs):
...
+                    raise xr.XuleRuleSetError('Need instance to process rules')
+
+        global _saved_taxonomies
+        used_taxonomies = process_xule(rule_set, modelXbrl, cntlr, options, _saved_taxonomies)
+        # Save one loaded taxonomy
+        new_taxonomy_keys = used_taxonomies.keys() - _saved_taxonomies.keys()
+        if len(new_taxonomy_keys) > 0:
+            for tax_key in list(new_taxonomy_keys)[:2]: # This take at most 2 taxonomies.
+                tax_key = next(iter(new_taxonomy_keys)) # randomly get one key
+                _saved_taxonomies[tax_key] = used_taxonomies[tax_key]
```
Comment (sagesmith-wf) [inline plugin/xule/__init__.py]: Do you need to check that `used_taxonomies[tax_key]` is not `None` and that it exists with a `.get()`?

File: plugin/xule/__init__.py (hunk 2 above)
Comment (derekgengenbacher-wf) [reply plugin/xule/__init__.py]: I do not understand this suggestion?

File: plugin/xule/__init__.py (hunk 1 above)
Comment (derekgengenbacher-wf) [reply plugin/xule/__init__.py]: You are not wrong, but the scope of this ticket was to fix the rule set loading error.

File: plugin/xule/__init__.py (hunk 3 above)
Comment (derekgengenbacher-wf) [reply plugin/xule/__init__.py]: Im not positive, but I would assume that XULE would always want to be multi threaded even if the xule_multi variable isnt set.

File: plugin/xule/__init__.py (hunk 4 above)
Comment (derekgengenbacher-wf) [reply plugin/xule/__init__.py]: You are not wrong, but the scope of this ticket was to fix the rule set loading error.

**Comment by derekgengenbacher-wf** [comment]
@davidtauriello Just wanted to check in and see where we are at with this PR getting merged?

File: plugin/xule/__init__.py (hunk 3 above)
Comment (phillipengel) [reply plugin/xule/__init__.py:778]: This should be switched back to ``False`` for the default for the xule_muli option so it reads:

``if getattr(options, "xule_multi", False) and getattr(cntlr, "rule_set", None) is not None:``

The --xule-multi option is only used when running xule in a server. It is not multi threading, but multi processing. For xule, this requires running xule in the xule server mode with the Xule_Multi module. The Xule_Multi module is not part of the DQC distribution. It signals the server to create multiple copies of Arelle in different processes and do some preloading of calculated constants. The server than can split the work between the processes. Its basically a poor man version of multi threading. But it requires some updates to Arelle itself, so it doesn't run with the official distribution of Arelle.

In normal running, the rule set is not supplied, so this IF statement will evaluate to False and execute the ELSE clauses. So it probably doesn't affect the normal run. But leaving this defaulting to True here could cause some unexpected issues. when running with a supplied rule set.

@davidtauriello will switch this back to False after PR #457 is merged.

**Comment by davidtauriello** [comment]
Added to v12 and modified as part of #471 public exposure

**Pull request closed without merging**